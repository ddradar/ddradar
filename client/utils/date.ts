export function unixTimeToString(unixTime: number) {
  const d = new Date(unixTime * 1000)
  return d.toLocaleString()
}
