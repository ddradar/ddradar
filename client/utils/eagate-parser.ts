import { getDanceLevel, UserScore } from '~/api/score'

type ChartScore = Omit<UserScore, 'userId' | 'userName'>

/**
 * Convert music data to { songId: Score[] } Record.
 * - https://p.eagate.573.jp/game/ddr/ddra20/p/playdata/music_data_single.html
 * - https://p.eagate.573.jp/game/ddr/ddra20/p/playdata/music_data_double.html
 */
export function musicDataToScoreList(
  sourceCode: string
): Record<string, ChartScore[]> {
  const doc = new DOMParser().parseFromString(sourceCode, 'text/html')
  const dataTable = doc.getElementById('data_tbl')
  if (!dataTable) throw new Error('invalid html')

  // Get PlayStyle from columns count
  const headerColumns =
    dataTable
      .getElementsByClassName('column')[0]
      ?.getElementsByClassName('rank')?.length ?? 0
  if (headerColumns !== 4 && headerColumns !== 5)
    throw new Error('invalid html')
  const playStyle = headerColumns === 5 ? 1 : 2

  const result: Record<string, ChartScore[]> = {}

  const songs = dataTable.getElementsByClassName('data')
  for (let i = 0; i < songs.length; i++) {
    const songRow = songs[i]

    // Get songName & songId from first column
    const songCol = songRow
      .getElementsByTagName('td')[0]
      .getElementsByClassName('music_info')[0]
    const songId = songCol
      .getAttribute('href')
      ?.replace(
        /^\/game\/ddr\/ddra20\/p\/playdata\/music_detail.html\?index=([01689bdiloqDIOPQ]{32})$/,
        '$1'
      )
    const songName = songCol.textContent?.trim()
    if (!songId || !songName) continue

    result[songId] = []

    const charts = songRow.getElementsByClassName('rank')
    for (let j = 0; j < charts.length; j++) {
      const chart = charts[j]

      // Get difficulty
      const difficulty = getDifficulty(chart.id.toLowerCase())

      // Get score
      const scoreText =
        chart.getElementsByClassName('data_score')[0].textContent ?? ''
      const score = parseInt(scoreText, 10)
      if (!Number.isInteger(score) || score < 0 || score > 1000000) continue

      // Get clearLamp
      const fcImageUrl = chart
        .getElementsByClassName('music_info')[0]
        .getElementsByTagName('img')[1]
        .src.toLowerCase()
      const rankImageUrl = chart
        .getElementsByClassName('music_info')[0]
        .getElementsByTagName('img')[0]
        .src.toLowerCase()
      const isFailed = /^.+\/rank_s_e\.png$/.test(rankImageUrl)
      const clearLamp =
        getClearLamp(fcImageUrl) ?? (isFailed ? 0 : score === 0 ? 1 : 2)

      // Get rank
      const rank = isFailed ? 'E' : getDanceLevel(score)

      result[songId].push({
        songId,
        songName,
        playStyle,
        difficulty,
        score,
        clearLamp,
        rank,
      })
    }
  }
  return result

  /** Get chart difficulty from element id */
  function getDifficulty(elementId: string) {
    if (elementId === 'beginner') return 0
    if (elementId === 'basic') return 1
    if (elementId === 'difficult') return 2
    if (elementId === 'expert') return 3
    if (elementId === 'challenge') return 4
    throw new Error('invalid html')
  }

  /** Get ClearLamp from full combo image URL */
  function getClearLamp(imageUrl: string) {
    // https://p.eagate.573.jp/game/ddr/ddra20/p/images/play_data/full_mar.png
    if (/^.+\/full_mar\.png$/.test(imageUrl)) return 7
    // https://p.eagate.573.jp/game/ddr/ddra20/p/images/play_data/full_perfect.png
    if (/^.+\/full_perfect\.png$/.test(imageUrl)) return 6
    // https://p.eagate.573.jp/game/ddr/ddra20/p/images/play_data/full_great.png
    if (/^.+\/full_great\.png$/.test(imageUrl)) return 5
    // https://p.eagate.573.jp/game/ddr/ddra20/p/images/play_data/full_good.png
    if (/^.+\/full_good\.png$/.test(imageUrl)) return 4
    // https://p.eagate.573.jp/game/ddr/ddra20/p/images/play_data/full_none.png
    return null
  }
}
