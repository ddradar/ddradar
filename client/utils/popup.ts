import type { BuefyNamespace } from 'buefy/types'

const option = { position: 'is-top', hasIcon: true } as const

export function success(
  $buefy: Pick<BuefyNamespace, 'notification'>,
  message: string
) {
  $buefy.notification.open({ message, type: 'is-success', ...option })
}

export function warning(
  $buefy: Pick<BuefyNamespace, 'notification'>,
  message: string
) {
  $buefy.notification.open({ message, type: 'is-warning', ...option })
}

export function danger(
  $buefy: Pick<BuefyNamespace, 'notification'>,
  message: string
) {
  $buefy.notification.open({ message, type: 'is-danger', ...option })
}
