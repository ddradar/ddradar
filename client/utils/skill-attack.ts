import { getDanceLevel, UserScore } from '~/api/score'

type MusicScore = Omit<UserScore, 'songId' | 'userId' | 'userName' | 'level'>

/** Read shift-jis encoded text file asynchronously */
export function readTextAsync(file: File) {
  return new Promise<string>((resolve, reject) => {
    const reader = new FileReader()
    reader.onload = () => resolve(reader.result as string)
    reader.onerror = () => reject(reader.error)
    reader.readAsText(file, 'shift-jis')
  })
}

/**
 * Convert Skill Attack score list to ScoreList
 * - http://skillattack.com/sa4/data/dancer/{ddrCode}/score_{ddrCode}.txt
 */
export function scoreTexttoScoreList(text: string) {
  const rows = text.trim().split('\n')
  const result: Record<string, MusicScore[]> = {}
  rows.forEach(r => {
    /**
     * - [0]: skillAttackId
     * - [1]: 0:SP, 1:DP
     * - [2]: 0:BEGINNER, 1:BASIC, ..., 4:CHALLENGE
     * - [3]: DDR CODE
     * - [4]: UNIX Time
     * - [5]: score
     * - [6]: 0:Played(Failed,Assisted,Clear), 1:FC(include GreatFC), 2:PFC, 3:MFC
     * - [7]: Song Name (HTML escaped)
     * - [8]: blank
     */
    const cols = r.split('\t')
    if (cols.length < 8) return

    const skillAttackId = cols[0]
    const playStyle = (parseInt(cols[1], 10) + 1) as 1 | 2
    const difficulty = parseInt(cols[2], 10) as 0 | 1 | 2 | 3 | 4
    const score = parseInt(cols[5], 10)
    const clearLamp =
      cols[6] === '3' ? 7 : cols[6] === '2' ? 6 : cols[6] === '1' ? 4 : 2
    const songName = unescapeHTML(cols[7])!

    if (result[skillAttackId] === undefined) result[skillAttackId] = []

    const scores = result[skillAttackId]
    const oldScore = scores.find(
      s => s.playStyle === playStyle && s.difficulty === difficulty
    )
    if (oldScore === undefined) {
      scores.push({
        songName,
        playStyle,
        difficulty,
        clearLamp,
        score,
        rank: getDanceLevel(score),
      })
    } else {
      oldScore.clearLamp = clearLamp
      oldScore.score = score
      oldScore.rank = getDanceLevel(score)
    }
  })
  return result

  function unescapeHTML(escaped: string) {
    const doc = new DOMParser().parseFromString(escaped, 'text/html')
    return doc.documentElement.textContent
  }
}
