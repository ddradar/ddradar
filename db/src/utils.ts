import type { JSONValue, SqlParameter } from '@azure/cosmos'

type Paths<T, Separator extends string = '.'> = T extends object
  ? {
      [K in keyof T]: `${Exclude<K, symbol>}${'' | `${Separator}${Paths<T[K], Separator>}`}`
    }[keyof T]
  : never

type Leaves<T> = T extends object
  ? {
      [K in keyof T]: `${Exclude<K, symbol>}${Leaves<T[K]> extends never ? '' : `/${Leaves<T[K]>}`}`
    }[keyof T]
  : never
export type ContainerPath<T> = `/${Leaves<T>}`

export type Column<TSchema, TModel = TSchema, Alias extends string = 'c'> =
  | `${Alias}.${Paths<TSchema> & Paths<TModel>}`
  | `${Alias}.${Exclude<Paths<TSchema>, Paths<TModel>>} AS ${Paths<TModel>}`

export type FuncColumn<
  DBSchema,
  T = DBSchema,
  Alias extends string = 'c',
> = `${string}${Alias}.${Extract<Exclude<keyof DBSchema, keyof T>, string>}${string} AS ${Extract<keyof T, string>}`

/** SQL WHERE condition */
export type QueryFilter<T, Alias extends string = 'c'> =
  | {
      /** WHERE condition */
      condition: `${Exclude<string, '@'>}${Alias}.${Extract<keyof T, string>}${Exclude<string, '@'>}`
      value?: never
    }
  | {
      /**
       * WHERE condition.
       * `"@"` replaces to `"@paramXX"`.
       */
      condition:
        | `${string}${Alias}.${Extract<keyof T, string>}${string}@${string}`
        | `${string}@${string}${Alias}.${Extract<keyof T, string>}${string}`
      /** Parameter value */ value: JSONValue
    }

export function generateQueryConditions(
  conditions: readonly { condition: string; value?: JSONValue }[]
) {
  return {
    queryConditions: conditions
      .map((c, i) => `(${c.condition.replace('@', `@param${i}`)})`)
      .join(' AND '),
    parameters: conditions
      .map((c, i) => ({ name: `@param${i}`, value: c.value }))
      .filter((c): c is SqlParameter => c.value !== undefined),
  }
}
