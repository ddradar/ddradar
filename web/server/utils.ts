import type { QueryObject } from 'ufo'

/**
 * Get string value from query object.
 * @param query Query object (returned by `useQuery(event)`)
 * @param key Query key
 * @returns Query value (if multiple, returns last value)
 */
export function getQueryString(query: QueryObject, key: string) {
  const value = query[key]
  return Array.isArray(value) ? value[value.length - 1] : value
}

/**
 * Get integer value from query object.
 * @param query Query object (returned by `useQuery(event)`)
 * @param key Query key
 * @returns Integer value or NaN
 */
export function getQueryInteger(query: QueryObject, key: string) {
  const value = getQueryString(query, key)
  if (value === undefined) return NaN
  const num = parseFloat(value)
  return Number.isInteger(num) ? num : NaN
}
