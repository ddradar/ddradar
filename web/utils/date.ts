export function unixTimeToString(unixTime: number): string {
  return new Date(unixTime * 1000).toLocaleString()
}
