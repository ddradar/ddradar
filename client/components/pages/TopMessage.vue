<template>
  <section>
    <b-message
      :type="type"
      :title="title"
      has-icon
      :icon="icon"
      aria-close-label="Close message"
    >
      <!-- eslint-disable-next-line vue/no-v-html -->
      <div class="content" v-html="markedContent" />
      <div class="content is-italic has-text-weight-light has-text-right">
        {{ date }}
      </div>
    </b-message>
  </section>
</template>

<script lang="ts">
import marked from 'marked'
import { Component, Prop, Vue } from 'nuxt-property-decorator'

import { unixTimeToString } from '~/utils/date'

@Component
export default class TopMessageComponent extends Vue {
  @Prop({ required: true, type: String })
  readonly type!: string

  @Prop({ required: false, type: String, default: null })
  readonly icon!: string | null

  @Prop({ required: true, type: String })
  readonly title!: string

  @Prop({ required: true, type: String })
  readonly body!: string

  @Prop({ required: true, type: Number })
  readonly time!: number

  get date() {
    return unixTimeToString(this.time)
  }

  get markedContent() {
    return marked(this.body)
  }
}
</script>

<style scoped>
section {
  padding: 0.75rem 0.75rem 0 0.75rem;
}
</style>
