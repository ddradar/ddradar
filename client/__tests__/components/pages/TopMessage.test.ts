import { createLocalVue, mount } from '@vue/test-utils'
import Buefy from 'buefy'
import { mocked } from 'ts-jest/utils'

import TopMessage from '~/components/pages/TopMessage.vue'
import { unixTimeToString } from '~/utils/date'

jest.mock('~/utils/date')
mocked(unixTimeToString).mockImplementation(unixTime =>
  new Date(unixTime * 1000).toUTCString()
)

const localVue = createLocalVue()
localVue.use(Buefy)

describe('/components/pages/TopMessage.vue', () => {
  /** 2020/8/13 0:00 (UTC) */
  const time = 1597276800
  const propsData = {
    type: 'is-info',
    title: 'Title',
    body: 'Message Body\n\n- First\n- Second',
    time,
  } as const

  describe('snapshot test', () => {
    test('renders default icon if not set', () => {
      // Arrange
      const wrapper = mount(TopMessage, {
        localVue,
        propsData: { ...propsData },
      })

      // Act - Assert
      expect(wrapper).toMatchSnapshot()
    })
    test('renders specified icon', () => {
      // Arrange
      const wrapper = mount(TopMessage, {
        localVue,
        propsData: { ...propsData, icon: 'account' },
      })

      // Act - Assert
      expect(wrapper).toMatchSnapshot()
    })
  })
})
