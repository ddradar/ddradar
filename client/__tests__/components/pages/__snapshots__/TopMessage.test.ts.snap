// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/components/pages/TopMessage.vue snapshot test renders default icon if not set 1`] = `
<section>
  <transition-stub name="fade">
    <article class="message is-info">
      <header class="message-header">
        <p>Title</p> <button type="button" aria-label="Close message" class="delete"></button>
      </header>
      <section class="message-body">
        <div class="media">
          <div class="media-left"><span class="icon is-large is-info"><i class="mdi mdi-information mdi-48px"></i></span></div>
          <div class="media-content">
            <p>Message Body</p>
            <div class="content is-italic has-text-weight-light has-text-right">
              2020/8/13
            </div>
          </div>
        </div>
      </section>
    </article>
  </transition-stub>
</section>
`;

exports[`/components/pages/TopMessage.vue snapshot test renders specified icon 1`] = `
<section>
  <transition-stub name="fade">
    <article class="message is-info">
      <header class="message-header">
        <p>Title</p> <button type="button" aria-label="Close message" class="delete"></button>
      </header>
      <section class="message-body">
        <div class="media">
          <div class="media-left"><span class="icon is-large is-info"><i class="mdi mdi-account mdi-48px"></i></span></div>
          <div class="media-content">
            <p>Message Body</p>
            <div class="content is-italic has-text-weight-light has-text-right">
              2020/8/13
            </div>
          </div>
        </div>
      </section>
    </article>
  </transition-stub>
</section>
`;
